src/shogun/mathematics/graph/ops/abstract/ShogunStorage.h

```diff
@@ -0,0 +1,398 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Gil Hoben
+ */
+
+#ifndef SHOGUN_OPERATOR_STORAGE_H_
+#define SHOGUN_OPERATOR_STORAGE_H_
+
+#include <shogun/lib/memory.h>
+#include <shogun/mathematics/graph/Shape.h>
+#include <shogun/mathematics/graph/Types.h>
+#include <shogun/util/zip_iterator.h>
+
+#include <numeric>
+#include <utility>
+
+namespace shogun
+{
+	namespace graph
+	{
+		namespace op
+		{
+			class ReshapeShogun;
+		}
+		/* A storage abstraction
+		 * The user should never see this. This is purely for
+		 * the backend to allocate storage. The user should use
+		 * the tensor class exclusively, in order to not mess
+		 * up shapes and so on.
+		 *
+		 */
+		class ShogunStorage
+		{
+			/* Managed storage. It's only job is to keep track of memory
+			 * and how to delete and allocate it.
+			 * It's unaware of what it actually owns. This is pretty much
+			 * a wrapper around std::shared_ptr for aligned memory.
+			 * Should this be thread safe?
+			 */
+			class InternalStorage
+			{
+				struct Transfer
+				{
+				};
+				struct NonOwning
+				{
+				};
+
+				using deleted_unique_ptr = std::unique_ptr<void,std::function<void(void*)>>;
```

Comment:
@vigsterkr not sure if storage should just be unique? For now it works, and if we ever need to make it accessible from multiple threads then we can makes shared_ptr and use the atomic overloads